### PR TITLE
[REVIEW] Change String typecasting to be inline with Pandas

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #2707 
- [x] Add cython bindings.
- [x] Make changes to `astype` logic.
- [x] Add docs.

Fixes: #5136

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/tests/test_dataframe.py [hunk 1]
```diff
@@ -3371,8 +3371,8 @@ def test_series_astype_null_cases():
 
     # string to other
     assert_eq(
-        gd.Series([1, 2, None, 3], dtype="int32"),
-        gd.Series(["1", "2", None, "3"]).astype("int32"),
+        gd.Series([1, 2, 3], dtype="int32"),
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/tests/test_dataframe.py:3374]: May be need to move to other test-case body as this test case is meant for null cases.

File: python/cudf/cudf/tests/test_sorting.py [hunk 2]
```diff
@@ -269,7 +269,7 @@ def test_dataframe_scatter_by_map(map_size, nelem, keep):
     df["a"] = np.random.choice(strlist[:map_size], nelem)
     df["b"] = np.random.uniform(low=0, high=map_size, size=nelem)
     df["c"] = np.random.randint(map_size, size=nelem)
-    df["d"] = df["a"]._column.as_categorical_column(np.int32)
+    df["d"] = np.zeros(nelem, dtype=np.int32)
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/tests/test_sorting.py:272]: Isn't it possible to keep a categorical column?

File: python/cudf/cudf/tests/test_string.py [hunk 3]
```diff
@@ -1828,3 +1828,36 @@ def test_string_str_decode_url(data):
     got = gs.str.url_decode()
     expected = pd.Series([urllib.parse.unquote(url) for url in data])
     assert_eq(expected, got)
+
+
+@pytest.mark.parametrize(
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/tests/test_string.py:1833]: Better to have two different set of test cases, one which fails for sure and one which passes for sure. Here, there is possibility that, something working might start giving ValueError later and this test case would fail to detect it.

File: python/cudf/cudf/core/column/string.py [hunk 4]
```diff
@@ -2035,6 +2039,22 @@ def as_numerical_column(self, dtype, **kwargs):
                     kwargs.update(format=fmt)
         kwargs.update(dtype=out_dtype)
 
+        if str_dtype.kind in ("i"):
+            if not cpp_all_integers(self):
+                bool_check = cpp_is_integer(self)
+                index = bool_check.astype("int8").find_first_value(0)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:2045]: I think we have `.any()` and `.all()` so you can check for all True vs all False.

File: python/cudf/cudf/tests/test_sorting.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_sorting.py:272]: This was previously creating a column of type `int32`. As having non-numerics used to give 0's in the result when there was a type-cast and now the same scenario will error, changed the column to directly initialize with 0's.

Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_sorting.py:272]: Earlier the test had a categorical column and now it has numerical column, so we are missing to test categorical. You can create a categorical column with only integers to test this.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py]: Removed this test as it is no more relevant. Instead testing with and without null separately in the newly added test.

File: python/cudf/cudf/tests/test_sorting.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_sorting.py]: Added the categorical back 👍

File: python/cudf/cudf/tests/test_dataframe.py [hunk 5]
```diff
@@ -4157,8 +4151,8 @@ def test_df_astype_string_to_other(as_dtype):
     expect["foo"] = expect_data
     expect["bar"] = expect_data
 
-    got = gdf.astype(as_dtype, **kwargs)
-    assert_eq(expect, got)
+    got = gdf.dropna().astype(as_dtype, **kwargs)
+    assert_eq(expect.dropna(), got)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:4155]: Why do we need to dropna here?

File: python/cudf/cudf/core/column/string.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Had a discussion offline with @kkraus14 and decided to not have the first error value(non-float/non-int) and instead display a generic error message that we are not able to type-cast due to invalid int/float.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 5 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py]: This was to support recent change where we error incase of `None` present in a series of type string and the typecast is to `int` types. Instead of dropping na, removed `None` from the data when typecast happens to `int32`.

File: python/cudf/cudf/tests/test_dataframe.py [hunk 6]
```diff
@@ -3378,12 +3378,6 @@ def test_series_astype_null_cases():
         ),
     )
 
-    # string to other
-    assert_eq(
-        gd.Series([1, 2, None, 3], dtype="int32"),
-        gd.Series(["1", "2", None, "3"]).astype("int32"),
-    )
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:3385]: This should still work, why is this removed?

File: python/cudf/cudf/tests/test_dataframe.py [hunk 7]
```diff
@@ -4136,7 +4130,7 @@ def test_df_astype_string_to_other(as_dtype):
         data = ["2001-01-01", "2002-02-02", "2000-01-05", None]
         kwargs = {"format": "%Y-%m-%d"}
     elif as_dtype == "int32":
-        data = [1, 2, 3, None]
+        data = [1, 2, 3]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:4133]: Can we make the float explicitly `NaN` and datetime explicitly `NaT` to be more clear on the expected behavior? I.E. this behavior may implicitly change in Pandas 1.0+ with `pd.NA`.

File: python/cudf/cudf/tests/test_string.py [hunk 8]
```diff
@@ -1828,3 +1828,56 @@ def test_string_str_decode_url(data):
...
+        (["++++1", "--2"], "int"),
+    ],
+)
+@pytest.mark.parametrize("obj_type", [None, "str", "category"])
+def test_string_typecast_error(data, obj_type, dtype):
+    psr = pd.Series(data, dtype=obj_type)
+    gsr = Series(data, dtype=obj_type)
+
+    with pytest.raises(ValueError):
+        psr.astype(dtype=dtype)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:1880]: Suggest capturing the type of error from here and ensuring that we raise the same error in the `gsr.astype` call.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 6 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:3385]: Right now with present set of changes(which bring pandas behavior) will error in this case. 
We are trying to type-cast a `None` to int32 which is not supported in pandas. For example:

```python
>>> s = pd.Series(["10",None])
>>> s
0      10
1    None
dtype: object
>>> s.astype('int')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/generic.py", line 5882, in astype
    dtype=dtype, copy=copy, errors=errors, **kwargs
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 581, in astype
    return self.apply("astype", dtype=dtype, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 438, in apply
    applied = getattr(b, f)(**kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 559, in astype
    return self._astype(dtype, copy=copy, errors=errors, values=values, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 643, in _astype
    values = astype_nansafe(vals1d, dtype, copy=True, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/cast.py", line 707, in astype_nansafe
    return lib.astype_intsafe(arr.ravel(), dtype).reshape(arr.shape)
  File "pandas/_libs/lib.pyx", line 547, in pandas._libs.lib.astype_intsafe
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

So do we want to support type-casting a `None`/`null` in string to int typecast?

Comment (kkraus14) [reply python/cudf/cudf/tests/test_dataframe.py:3385]: Yes but in Pandas that's because they don't have nullable integers (in their standard type). For us we should be able to handle string `null` --> int `null`.

File: python/cudf/cudf/tests/test_string.py (hunk 8 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_string.py:1880]: By `type` you mean error "message"? `psr` Raises a `ValueError` for invalid literals.

Comment (kkraus14) [reply python/cudf/cudf/tests/test_string.py:1880]: I mean something like this:
```
try:
    psr.astype(dtype=dtype)
except as e:
    exception_type = type(e)

with pytest.raises(exception_type):
    gsr.astype(dtype=dtype)
```

File: python/cudf/cudf/tests/test_dataframe.py (hunk 7 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:4141]: This has been modified, except DateTime for which there seems to be an issue: https://github.com/rapidsai/cudf/issues/5117

File: python/cudf/cudf/core/column/string.py [hunk 9]
```diff
@@ -2038,7 +2038,12 @@ def as_numerical_column(self, dtype, **kwargs):
         kwargs.update(dtype=out_dtype)
 
         if str_dtype.kind in ("i"):
-            if not cpp_all_integers(self):
+            if self.null_count > 0:
+                string_col = self.dropna()
+            else:
+                string_col = self
+
+            if not cpp_all_integers(string_col):
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py]: It would be great if we moved this down to a parameter for the libcudf API. I imagine we could return True for nulls instead of False relatively easily and it would save us having to effectively do a stream compaction here. cc @davidwendt 

Not a blocker for merging though for sure.

File: python/cudf/cudf/tests/test_string.py [hunk 10]
```diff
@@ -1828,3 +1828,59 @@ def test_string_str_decode_url(data):
...
+    ],
+)
+@pytest.mark.parametrize("obj_type", [None, "str", "category"])
+def test_string_typecast_error(data, obj_type, dtype):
+    psr = pd.Series(data, dtype=obj_type)
+    gsr = Series(data, dtype=obj_type)
+
+    try:
+        psr.astype(dtype=dtype)
+        raise TypeError("Was expecting `psr.astype` to fail")
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:1881]: Does this break out of the try except or does it get caught by the `except` block below?

File: python/cudf/cudf/core/column/string.py (hunk 9 above)
Comment (davidwendt) [reply python/cudf/cudf/core/column/string.py]: Adding parameter to `cudf::strings::all_integers()` to have it return true for nulls would not be difficult.

Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Sure, in that case, should we hold off this PR until those changes are in, or should we adapt to the changes in a follow-up PR?

@davidwendt, we'll also need the same parameter for `all_floats` API too as we are handling the null case in a similar fashion too.

File: python/cudf/cudf/tests/test_string.py (hunk 10 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_string.py]: Yikes, that would have got caught up in `except` as well. Branched it out.

File: python/cudf/cudf/core/column/string.py (hunk 9 above)
Comment (kkraus14) [reply python/cudf/cudf/core/column/string.py]: No need to hold off, @galipremsagar can you write up an issue for adding a parameter to `all_integers()` / `all_floats()`?

Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Yep, link to the issue: https://github.com/rapidsai/cudf/issues/5130

File: python/cudf/cudf/core/series.py [hunk 11]
```diff
@@ -1652,7 +1652,16 @@ def label_encoding(self, cats, dtype=None, na_sentinel=-1):
...
-        cats = Series(cats).astype(self.dtype)
+        cats = Series(cats)
+        try:
+            # Where there is a type-cast from string to numeric types,
+            # there is a possibility for ValueError when strings
+            # are having non-numeric values, hence we will have
+            # to catch the exception and fill in with zeros.
+            cats = cats.astype(self.dtype)
+        except ValueError:
+            cats = Series(cupy.zeros(len(cats), dtype=self.dtype))
```
Comment (kkraus14) [inline python/cudf/cudf/core/series.py:1663]: What does a categories of all zeros even mean? If we can't safely cast what's passed into `cats` to the dtype of what we're encoding we should probably just throw an informative error, no?

Comment (galipremsagar) [reply python/cudf/cudf/core/series.py]: Cats with all zeros would result in all nulls for the `value` column on merge. We don't actually error but instead return `na_sentinel` after performing a merge against 0's. This was previous behavior.

But upon going through different possible categorical input combinations I realized that we can actually short-circuit the whole merging & sorting if there is a `ValueError` in type-cast, we just return a column filled with `na_sentinel` values.

File: python/cudf/cudf/core/series.py [hunk 12]
```diff
@@ -1652,19 +1652,42 @@ def label_encoding(self, cats, dtype=None, na_sentinel=-1):
...
+                name=None,
+            )
+
         order = Series(cupy.arange(len(self)))
         codes = Series(cupy.arange(len(cats), dtype=dtype))
 
         value = DataFrame({"value": cats, "code": codes})
         codes = DataFrame(
-            {"value": self.copy(deep=False)._data.columns[0], "order": order}
+            {"value": self._data.columns[0].copy(deep=False), "order": order}
```
Comment (kkraus14) [inline python/cudf/cudf/core/series.py:1678]: Should we make `order` just a column as well instead of a Series above?

File: python/cudf/cudf/core/series.py [hunk 13]
```diff
@@ -1652,19 +1652,42 @@ def label_encoding(self, cats, dtype=None, na_sentinel=-1):
...
         codes = codes.merge(value, on="value", how="left")
-        codes = codes.sort_values("order")["code"].fillna(na_sentinel)
 
-        cats.name = None  # because it was mutated above
+        # If the `code` column is full of nulls
+        # there is no need of sorting by `order`
+        if codes._data["code"].null_count == len(codes):
+            codes = codes["code"]
+        else:
+            codes = codes.sort_values("order")["code"]
```
Comment (kkraus14) [inline python/cudf/cudf/core/series.py:1687]: What's the motivation for adding this check?

Comment (galipremsagar) [reply python/cudf/cudf/core/series.py:1687]: So this check is to avoid sorting by `order` when all of `code` column is `null`. The reason `code` column can all be null here is if we have non-intersecting categories in `self` & the categories passed(`cats`).
example:
```python
>>> s = cudf.Series([1,2,3])
>>> t = cudf.CategoricalDtype([5,6,7])
>>> s.astype(t)
> /conda/envs/cudf/lib/python3.7/site-packages/cudf/core/series.py(1685)label_encoding()
-> if codes._data["code"].null_count == len(codes):
(Pdb) codes
   value  order  code
0      1      0  null
1      2      1  null
2      3      2  null
```

Comment (kkraus14) [reply python/cudf/cudf/core/series.py:1687]: Is this a common operation that we actually need the short circuiting for? Otherwise this feels like a premature optimization to me.

Comment (galipremsagar) [reply python/cudf/cudf/core/series.py]: Had a discussion offline with @kkraus14 and decided to revert this change since the scenario is not a common one.
